orc8r/cloud/go/service/middleware/unary/identity_decorator.go

```diff
@@ -102,82 +101,113 @@ func SetIdentityFromContext(ctx context.Context, _ interface{}, info *grpc.Unary
+		if err == nil {
+			return newCtx, newReq, resp, nil
 		}
+	} else if _, ok = ctxMetadata[CLIENT_CERT_CN_KEY]; ok {
+		// CN header is present while SN header is missing - possible
+		// security hack, either both or neither of the headers should be
+		// set
+		err = unexpectedCCN(ctxMetadata)
 	}
 
-	if info != nil {
-		// Check if the call is for a allowlisted method - anything is allowed
-		// do this check past possible identity decoration to still allow to add
-		// valid identity even to allowlisted requests
-		if _, ok := identityDecoratorBypassList[info.FullMethod]; ok {
-			// Bypass method (Bootstrapper & Co.), shortcut...
-			return newCtx, newReq, resp, nil
-		}
+	if isAllowlistedCall(info) {
+		return newCtx, newReq, resp, nil
 	}
 
-	// Only allow local clients if there is no previous errors (snok == false)
 	if err == nil {
 		// We assume that only external calls forwarded by cloud proxy (or unit
 		// tests) will have CSN & CCN headers set. The absence of the headers
 		// along with client IP verification will indicate a local service to
 		// service or Obsidian to service call
-		// For internal calls, no identity verification needed, just make sure
-		// it's a local client
-		err = ensureLocalPeer(ctx)
-		if err != nil {
-			var rpc string
-			if info != nil {
-				rpc = info.FullMethod
-			} else {
-				rpc = "Undefined"
-			}
-			glog.Infof("Empty CTX Metadata from non-local %s client: %v", rpc, err)
-		}
+		err = checkLocalClientCall(ctx, info)
 	}
 	return newCtx, newReq, resp, err
 }
 
+func noMetadata() error {
+	glog.Info(ERROR_MSG_NO_METADATA)
+	return status.Error(codes.Unauthenticated, ERROR_MSG_NO_METADATA)
+}
+
+func unexpectedCCN(ctxMetadata metadata.MD) error {
+	glog.Infof("CCN is present without SCN in metadata: %+v", ctxMetadata)
+	return status.Error(codes.Unauthenticated, "Inconsistent Request Signature")
+}
+
+func multipleCSN(ctxMetadata metadata.MD) error {
+	glog.Infof("Multiple CSNs found in metadata: %+v", ctxMetadata)
+	return status.Error(codes.Unauthenticated, "Multiple CSNs present")
+}
+
+func findIdentityFromCSNlist(snlist []string, ctx context.Context, ctxMetadata metadata.MD) (context.Context, error) {
+	if len(snlist) == 1 {
+		// One CSN is found, find Caller's Identity associated with it
+		newCtx, err := findIdentityFromUniqueCSN(snlist[0], ctx, ctxMetadata)
+		return newCtx, err
+	} else {
+		// there is a certificate serial number (CSN) list in CTX
+		// there can be only one CSN, error out if not
+		return nil, multipleCSN(ctxMetadata)
+	}
+}
+
+func findIdentityFromUniqueCSN(sn string, ctx context.Context, ctxMetadata metadata.MD) (newCtx context.Context, err error) {
+	// Check if SN is the reserved value used for all inter-orc8r calls
+	if sn == registry.ORC8R_CLIENT_CERT_VALUE {
+		return newCtx, nil
+	}
+	var gwIdentity *protos.Identity
+	var certExpTime int64
+	gwIdentity, certExpTime, err = findGatewayIdentity(ctx, sn, ctxMetadata)
```

Comment:
isn't that equivalent to:
```suggestion
	gwIdentity, certExpTime, err := findGatewayIdentity(ctx, sn, ctxMetadata)
```

Reply:
See comment below.